securedrop_client/gui/conversation/print/printer.py

```diff
@@ -0,0 +1,125 @@
+        return self._status
+
+    @status.setter
+    def status(self, new: Status) -> None:
+        if new != self._status:
+            self._status = new
+            self.status_changed.emit()
+
+    def start_and_watch_on(self, signal: pyqtSignal) -> None:
+        """Allows print queue to be started and its status to be watched regularly, in a thread-safe manner."""  # noqa: E501
+        signal.connect(self._printing_service._start_and_watch)
+
+    def stop_watching_on(self, signal: pyqtSignal) -> None:
+        """Allows to stop watching the print queue status in a thread-safe manner."""
+        signal.connect(self._printing_service._stop_watching)
+
+    def enqueue_job_on(self, signal: pyqtSignal) -> None:
+        signal.connect(self._printing_service._enqueue_job)
+
+    @pyqtSlot()
+    def _on_job_enqueued(self) -> None:
+        self.job_done.emit("unknown job ID")  # FIXME
+
+    @pyqtSlot(object)
+    def _on_job_enqueuing_failed(self, error: export.ExportError) -> None:
+        reason = str(error)  # FIXME check the output
+        self.job_failed.emit("unknown job ID", reason)  # FIXME
+
+    @pyqtSlot()
+    def _on_printer_found_ready(self) -> None:
+        self.status = self.StatusReady
+
+    @pyqtSlot(object)
+    def _on_printer_not_found_ready(self, reason: object) -> None:
+        self.status = self.StatusUnreachable
+
+    class _ExportServiceConnector(QObject):
+        """Connect a printer to the export service in a thread-safe way."""
+
+        _printer_start_requested = pyqtSignal()
+        _job_enqueuing_requested = pyqtSignal(list)
+
+        _DEFAULT_POLLING_DELAY_IN_MILLISECONDS = 3000
+
+        def __init__(self, printer: "Printer", service: export.Service) -> None:
+            super().__init__()
+
+            # Command signals. Signals are thread-safe, and this allows
+            # the service to run in a different thread. They represent events,
+            # but arguably are a constuction to send commands.
+            service.connect_signals(
+                print_preflight_check_requested=self._printer_start_requested,
+                print_requested=self._job_enqueuing_requested,
+            )
+
+            # Event signals. These signals are used in the canonical way,
+            # the service broadcasts events that the printer consumes.
+            service.printer_preflight_failure.connect(printer._on_printer_not_found_ready)
+            service.printer_preflight_success.connect(printer._on_printer_found_ready)
+            service.print_call_failure.connect(printer._on_job_enqueuing_failed)
+            service.print_call_success.connect(printer._on_job_enqueued)
+
+            # Service watch signals. Allow to poll the service,
+            # in order to check its status regularly.
+            service.print_call_failure.connect(self._maybe_schedule_printer_status_check)
+            service.print_call_success.connect(self._maybe_schedule_printer_status_check)
```

Comment:
Mistake: those are meant to be `printer_preflight_{success,failure}`. (That is: schedule the next check when the previous check is done, if polling is intended. Otherwise, do nothing.)